Refactor: unify refactoring and transformation execution api

balsa-sarenac opened the pull request:
This PR unifies refactoring/transformation API so that they can be used one with another. Changes are mainly in `RBTransformation` and include:

* removing `preconditionChecking` variable
* removing `delegatesPrecondition` for composite transformations
* checking `applicabilityConditions` by default and fixing bunch of tests to support this
* extracting `breakingChangeConditions` out of `applicabilityConditions` NOTE: all breaking change checks are __NOT__ called anymore, that's why so many failing tests need to be fixed. I'll fix them by splitting them on refactoring and transformation tests.

There are a bunch of failing tests, I am pushing this now since I won't be able to work on this until next week. Partial review is welcome.

balsa-sarenac committed:
Refactor: transformations check `applicabilityConditions` by default

balsa-sarenac committed:
Fix: `assertEmpty:` changes are tests for failed transformations

Since transformations didn't check preconditions by default these tests whould assert
that when transforamtion is badly configured it returns zero changes.

balsa-sarenac committed:
Refactor: remove `preconditionChecking` variable for transformations

Transformations will now always check applicability preconditions by default.

balsa-sarenac committed:
Refactor: remove remaining usages of `delegatesPreconditions`

balsa-sarenac committed:
Refactor:  remove-variable-transformation preconditions

Split preconditions  on applicability and breaking change changes

balsa-sarenac committed:
Fix: use existing category name

balsa-sarenac committed:
Refactor: extract breaking change conditions from applicability ones

balsa-sarenac committed:
Fix: add missing accessors to fix failing test

balsa-sarenac committed:
Refactor: extract breaking changes from applicability conditions for add-temp-var

balsa-sarenac committed:
Refactor: extract breaking changes from applicability conditions for add-variable

balsa-sarenac committed:
Refactor: split conditions for push-down-variable transformation

balsa-sarenac committed:
Refactor: remove unused `preconditionChecking` variable

jecisc commented on src/Refactoring-DataForTesting/RBTransformationRuleTestData.class.st:
```diff
@@ -280,6 +280,18 @@ RBTransformationRuleTestData >> checkMethod: aSmalllintContext [
 ]
 
+{ #category : #accessing }
+RBTransformationRuleTestData >> class [
+
+	^ class
+]
```
Usually it is a bad idea to have a method called "class" because it overrides the method allowing to get the class of an object.

For example, on compiled method. instead of having a method called "class" it was named "methodClass" so that we can still ask the class of the method.

balsa-sarenac replied:
Yes, I'm aware. I only did it because this is a test fixture method. This will be removed in the next commit.

balsa-sarenac committed:
Fix: avoid overriding `class` method

balsa-sarenac committed:
Fix: instance variable can be added to metaclass

balsa-sarenac committed:
Fix: variables can be removed from metaclass

balsa-sarenac committed:
Refactor: change test name to better reflect behaviour

balsa-sarenac committed:
Fix: class variables cannot be added to metaclass

balsa-sarenac committed:
Refactor: class variables are always defined on instance side

balsa-sarenac committed:
Refactor: extract part of breaking changes from applicability conditions

Other part will be refactored later. This one was blocking.
Extracting others requires refactoring remove-class to a driver and 
handling better the batch processing removal.

balsa-sarenac committed:
Docs: add class comment to abstract variables transformation

balsa-sarenac committed:
Fix: by default class variables are defined on instnace side, but their accessors on class side

This was implicit before, this makes it more explicit.
This will potentially be changed in future.

balsa-sarenac committed:
Fix: set instnace variable value when setter already exists

balsa-sarenac committed:
Refactor: rename methods that return selector so that its name reflects its return value

balsa-sarenac committed:
Fix: test is not expected to fail anymore since transformation don't check breaking changes

balsa-sarenac committed:
Fix: change expected changes size number

I've checked and all the changes it produced are aligned with transformations
that are being executed.

balsa-sarenac committed:
Refactor: split brekaing change and applicability preconditions

balsa-sarenac committed:
Refactor: separate transformation and refactoring tests for add-class-variable

Transformation tests are parametrized and used on both refactoring and transformation since
both of them should fail on applicability checks.
Brekaing change checks are tested only on refactorings now.

balsa-sarenac committed:
Refactor: let the defining class be the calss that was supplied

Subclasses can then change if they want instnace or class side on their own

balsa-sarenac committed:
Fix: unskip tests that now work thanks to previous changes

balsa-sarenac committed:
Refactor: split preconditions on applicability and breaking changes

balsa-sarenac committed:
Refactor: split testss for add-instance-variable on refactoring and transformation tests

balsa-sarenac committed:
Test: add tests that verify applicability conditions of transformation

balsa-sarenac committed:
Refactor: split refactoring tests for breaking changes from transformation for add-method

balsa-sarenac committed:
Fix: update expected number of transformation changes

Transformation actually fixes all the references so expected number
is wrong here. This will be reverted in the future when we move
breaking change transformations to refactoring.

balsa-sarenac committed:
Fix: add-temp that already exist is an applicability condition

balsa-sarenac committed:
Refactor: extract breaking change test from transformation test for push-down-class-var

balsa-sarenac committed:
Refactor: extract breaking change checks from transformation tests for remove-class-var

balsa-sarenac committed:
Fix: for class variables we need to incluce metaclass and subclasses for reference update

balsa-sarenac committed:
Refactor: skip adding temporary if it already exists in selected scope

balsa-sarenac committed:
Fix: remove test that checks if transformation fails if variable already exists

balsa-sarenac committed:
Merge branch 'Pharo12' into refactor/unify-refactoring-transformation-execution-api

balsa-sarenac committed:
Refactor: migrating extract-method to new transformation api
